### PR TITLE
Adding docs to README.md

### PR DESCRIPTION
**Pull request by jmencak**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,166 @@
...
 
-The Cluster Node Tuning Operator manages cluster node-level tuning for [OpenShift](https://openshift.io/).
+The Cluster Node Tuning Operator manages cluster node-level tuning for
+[OpenShift](https://openshift.io/).
+
+The majority of high-performance applications require some
+level of kernel tuning.  The operator provides a unified
+management interface to users of node-level sysctls and more
+flexibility to add [custom tuning](#custom-tuning-specification)
+specified by user needs.  The operator manages containerized
```
Comment (ahardin-rh) [inline README.md]: the containerized

File: README.md [hunk 2]
```diff
@@ -1,3 +1,166 @@
...
+[tuned](https://github.com/redhat-performance/tuned/)
+daemon for [OpenShift](https://openshift.io/) as
+a Kubernetes DaemonSet.  It ensures [custom tuning
+specification](#custom-tuning-specification) is passed to all
+[containerized tuned](https://github.com/openshift/openshift-tuned)
+daemons running in the cluster in the format they understand.
```
Comment (ahardin-rh) [inline README.md]: who is "they"? If it's the daemons, we can just say "the format that the daemons understand."

File: README.md [hunk 3]
```diff
@@ -1,3 +1,166 @@
...
+The daemons run on all nodes in the cluster, one per node.
+
+
+## Deploying the node tuning operator
+
+The operator is deployed by applying the manifests in the operator's
+`/manifests` directory.  It tries to be useful out of the box by
```
Comment (ahardin-rh) [inline README.md]: I would reword this sentence to say "It automatically creates a default deployment and configuration
([CR](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/))
for the tuned daemons" but then indicate what makes this a useful feature. Is it useful because it saves the user time, or adds value in another way?

File: README.md [hunk 4]
```diff
@@ -1,3 +1,166 @@
...
+
+```
+$ oc get Tuned
+NAME      AGE
+default   1h
+```
+
+## Custom tuning specification
+
+For an example of a tuning specification, please refer to
```
Comment (ahardin-rh) [inline README.md]: s/please refer to/refer to

File: README.md [hunk 5]
```diff
@@ -1,3 +1,166 @@
...
+`/assets/tuned/07-cr-tuned.yaml` in the operator's directory or to
+the resource created in a live cluster by:
+
+```
+oc get Tuned/default -o yaml
```
Comment (ahardin-rh) [inline README.md]: $ at the start of the command

File: README.md [hunk 6]
```diff
@@ -1,3 +1,166 @@
...
+```
+
+The custom resource for the operator has two major sections.  The first
+section `profile:` is a list of tuned profiles and their names, the second
```
Comment (ahardin-rh) [inline README.md]: , `profile:`, 

names. The second, `recommend:`,

File: README.md (hunk 3 above)
Comment (ahardin-rh) [reply README.md]: The first instance of "custom resource", I would spell out as "custom resource (CR)". Any instance afterwards can be just "CR"

File: README.md [hunk 7]
```diff
@@ -1,3 +1,166 @@
...
+`recommend:` defines the profile selection logic.
+
+Multiple custom tuning specifications can co-exist as multiple CRs
+in the operator's namespace.  The existence of new or deletion of
```
Comment (ahardin-rh) [inline README.md]: The existence of new CRs or the deletion of old CRs is detected by the operator. All existing custom tuning specifications are merged and appropriate objects for the containerized tuned daemons are updated.

File: README.md [hunk 8]
```diff
@@ -1,3 +1,166 @@
...
+      [main]
+      summary=Description of tuned_profile_n profile
+
+      # tuned_profile_n profile settings
+```
+
+
+### Recommended profiles
+
+The profile selection logic is defined by the `recommend:` section of the CR.
```
Comment (ahardin-rh) [inline README.md]: The `profile:` section

File: README.md [hunk 9]
```diff
@@ -1,3 +1,166 @@
...
+
+  # ...
+
+  - match:
+    <match>
+    priority: <priority>
+    profile: <tuned_profile_name>       # e.g. tuned_profile_n
+```
+
+If `<match>` is omitted, a profile match (i.e. _true_) is assumed.
```
Comment (ahardin-rh) [inline README.md]: s(/i.e. _true_)/for example, _true_)

File: README.md [hunk 10]
```diff
@@ -1,3 +1,166 @@
...
+`<match>` is an optional array recursively defined as follows:
+
+```
+    - label: <label_name>     # node or pod label name
+      value: <label_value>    # optional node or pod label value; if omitted, the presence of <label_name> is enough to match
+      type: <label_type>      # optional node or pod type ("node" or "pod"); if omitted, "node" is assumed
+      <match>                 # an optional <match> array
+```
+
+If `<match>` is not omitted all nested `<match>` sections must also
```
Comment (ahardin-rh) [inline README.md]: If `<match>` is not omitted, all nested `<match>` sections must also evaluate to _true_. Otherwise,  _false_ is assumed and the profile with the respective `<match>` section will not be applied or recommended. Therefore,

File: README.md [hunk 11]
```diff
@@ -1,3 +1,166 @@
...
+evaluate to _true_, otherwise _false_ is assumed and the profile with
+the respective `<match>` section will not be applied/recommended.
+Therefore the nesting (child `<match>` sections) works as logical
```
Comment (ahardin-rh) [inline README.md]: what does "logical" mean here? I only know "logical" as an adjective.

s/AND/and

File: README.md [hunk 12]
```diff
@@ -1,3 +1,166 @@
...
+AND operator.  Conversely, if any item of the `<match>` array matches,
+the entire `<match>` array evaluates to _true_, therefore the array
```
Comment (ahardin-rh) [inline README.md]: _true_. Therefore, the array acts as logical or operator.

File: README.md [hunk 13]
```diff
@@ -1,3 +1,166 @@
...
+    priority: 20
+    profile: openshift-control-plane
+  - priority: 30
+    profile: openshift-node
+```
+
+The CR above is translated for the containerized tuned daemon into
+its recommend.conf file based on the profile priorities.  The profile
+with the highest priority (10) is openshift-control-plane-es and
+therefore it is considered first.  The containerized tuned daemon
```
Comment (ahardin-rh) [inline README.md]: , therefore,

File: README.md [hunk 14]
```diff
@@ -1,3 +1,166 @@
...
+running on a given node looks to see if there is a pod running on the
+same node with tuned.openshift.io/elasticsearch label set.  If not,
```
Comment (ahardin-rh) [inline README.md]: node with the

File: README.md [hunk 15]
```diff
@@ -1,3 +1,166 @@
...
+the entire `<match>` section evaluates as _false_.  If there is such a
+pod with the label, in order for the `<match>` section to evaluate to
+_true_ the node label also needs to be node-role.kubernetes.io/master
```
Comment (ahardin-rh) [inline README.md]: , the node label

File: README.md [hunk 16]
```diff
@@ -1,3 +1,166 @@
...
+OR node-role.kubernetes.io/infra.
```
Comment (ahardin-rh) [inline README.md]: s/OR/or

File: README.md [hunk 17]
```diff
@@ -1,3 +1,166 @@
...
+
+If the labels for the profile with priority 10 matched,
+openshift-control-plane-es profile is applied and no other profile is 
+considered.  In case the node/pod label combination did not match,
```
Comment (ahardin-rh) [inline README.md]: s/In case/If

File: README.md [hunk 18]
```diff
@@ -1,3 +1,166 @@
...
+pod with the label, in order for the `<match>` section to evaluate to
+_true_ the node label also needs to be node-role.kubernetes.io/master
+OR node-role.kubernetes.io/infra.
+
+If the labels for the profile with priority 10 matched,
+openshift-control-plane-es profile is applied and no other profile is 
+considered.  In case the node/pod label combination did not match, 
+the second highest priority profile (openshift-control-plane) is considered.
+This profile is applied if the containerized tuned pod runs on a node with 
+labels node-role.kubernetes.io/master OR node-role.kubernetes.io/infra.
```
Comment (ahardin-rh) [inline README.md]: s/OR/or

File: README.md [hunk 19]
```diff
@@ -1,3 +1,166 @@
...
+
+Finally, the profile `openshift-node` has the lowest priority of 30.
+It lacks the `<match>` section and therefore will always match and
```
Comment (ahardin-rh) [inline README.md]: , therefore,

Comment (ahardin-rh) [reply README.md]: match. It acts as a profile catch-all to set the  `openshift-node` profile, if no

Comment (ahardin-rh) [reply README.md]: What do you mean by a catch-all?

File: README.md (hunk 9 above)
Comment (jmencak) [reply README.md]: Are you sure about this change?  This would the semantics.  If `<match>` is omitted then it automatically "matches", i.e. the match evaluates to "true".  "for example" sounds strange here to me.

Comment (ahardin-rh) [reply README.md]: @jmencak Oh okay, I see. We tend to avoid i.e. and e.g. in our docs, but if this statement is more correct as-is, then we can let it be.

File: README.md (hunk 11 above)
Comment (jmencak) [reply README.md]: https://en.wikipedia.org/wiki/Logical_connective

Comment (ahardin-rh) [reply README.md]: @jmencak I learn something new every day :rainbow:

File: README.md (hunk 19 above)
Comment (jmencak) [reply README.md]: To me that means something that ends up being used if nothing else "matches" in this case.  Last resort.  For example, if none of the node/pod labels allow for a higher priority profile to be set, the lowest priority (catch-all) profile is used.

Comment (ahardin-rh) [reply README.md]: @jmencak Okay, that makes more sense. Thanks!

**Comment by jmencak** [comment]
Thanks for the review @ahardin-rh.  It is late, _therefore_, it is bed time.

File: README.md (hunk 18 above)
Comment (jmencak) [reply README.md]: As for OR/AND, I wanted to make it more obvious that these are logical connectives.  Not sure what would be a better way.

Comment (jmencak) [reply README.md]: using _or_/_and_ for these

**Comment by jmencak** [comment]
Reviewer comments addressed.

/lgtm

**Comment by jeremyeder** [comment]
Thanks for the extensive review, @ahardin-rh.  I read it as well and don't have anything substantive to add.  Thanks for completing this, @jmencak.
